### PR TITLE
Fix setter usage for checkpoint io and precision in TTP

### LINKED ISSUE #11097: Lazy initialize properties in Strategies
## Proposed refactor
Strategy has four properties in init : Accelerator, Precision_plugin, Checkpoint_io and Cluster_enviroment.

User could pass in a strategy class or a strategy str into trainer.
- If user pass a string, accelerator_connector will initialize the strategy with the properties selection. 
- If user pass a strategy class, the logics get a bit mess. We should track what user has passed in V.S strategy default value V.S. Trainer.accelerator_connector 

This issue propose lazy initialization for strategy properties. @awaelchli and @ananthsub also bought this up before

### Motivation
For correctness, maintenance and enable future simplifications 

### Pitch

Current Training_type_plugin __init__ will set default value to precision_plugin and checkpoint_io:
https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/plugins/training_type/training_type_plugin.py#L40-L58
```python
    def __init__(
        self, checkpoint_io: Optional[CheckpointIO] = None, precision_plugin: Optional[PrecisionPlugin] = None
    ) -> None:
        self._model: Optional[Module] = None
        checkpoint_io = checkpoint_io if checkpoint_io is not None else TorchCheckpointIO()
        self._checkpoint_io = checkpoint_io
        self._precision_plugin = precision_plugin if precision_plugin is not None else PrecisionPlugin()

```

Proposal

```python
def __init__(
        self, checkpoint_io: Optional[CheckpointIO] = None, precision_plugin: Optional[PrecisionPlugin] = None
    ) -> None:
        self._model: Optional[Module] = None
        self._checkpoint_io = checkpoint_io
        self._precision_plugin = precision_plugin

@property
def checkpoint_io(self):
       return self._checkpoint_io if self._checkpoint_io is not None else TorchCheckpointIO()

@setter
def checkpoint_io(checkpoint_io):
       self._checkpoint_io = checkpoint_io

@property
def precision_plugin(self):
      return self._precision_plugin if self._precision_plugin is not None else PrecisionPlugin()

@setter
def precision_plugin(precision_plugin):
       self._precision_plugin = precision_plugin
```

In accelerator_connector.py, we know:
`training_type_plugin._precision_plugin` before we call setter is what user passed in, not the default value.

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @justusschock @awaelchli @akihironitta @kaushikb11 @ananthsub

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Fixes inconsistencies with the `precision_plugin` and `checkpoint_io` properties in the `TrainingTypePlugin`.

Fixes #11097 

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [n/a] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [n/a] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @justusschock @awaelchli @akihironitta @borda @jjenniferdai @kaushikb11 @ananthsub

### CODE REVIEW

File: pytorch_lightning/plugins/training_type/training_type_plugin.py
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:66]: This has been added for consistency. Ideally neither would have a setter but the `AcceleratorConnector` requires it.

File: pytorch_lightning/plugins/training_type/training_type_plugin.py
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:50]: The setters weren't getting used.

File: pytorch_lightning/plugins/training_type/deepspeed.py
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/deepspeed.py:880]: #8743 first implemented a `DeepSpeedCheckpointPlugin` which was removed before merge, so this is not necessary

File: pytorch_lightning/plugins/training_type/training_type_plugin.py
Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:50]: And how are they getting used now? There can't be a public attribute and a setter with the same name.

File: pytorch_lightning/plugins/training_type/training_type_plugin.py
Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:66]: How is this not creating an error? Something is not right here

File: pytorch_lightning/plugins/training_type/deepspeed.py
Comment (rohitgr7) [reply pytorch_lightning/plugins/training_type/deepspeed.py:880]: if someones does this:
```py
plugin = DeepSpeedPlugin()
plugin.checkpoint_io = CustomCheckpointIO()
```
this should still raise this exception?

File: pytorch_lightning/plugins/training_type/tpu_spawn.py
Comment (rohitgr7) [inline pytorch_lightning/plugins/training_type/tpu_spawn.py:345]: quick ques: are we forcing users to always subclass `XLACheckpointIO` and not `CheckpointIO` even if their implementation is correct?

File: pytorch_lightning/plugins/training_type/training_type_plugin.py
Comment (carmocca) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:50]: This will use the setter to set `self._precision_plugin`.

If we don't do it like this, any custom logic in the setter won't run

File: pytorch_lightning/plugins/training_type/training_type_plugin.py
Comment (carmocca) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:66]: Why would it?

File: pytorch_lightning/plugins/training_type/tpu_spawn.py
Comment (carmocca) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py:345]: It's un-pythonic to have these `isinstance` checks. I prefer to remove them, however, in this PR I'm trying to keep the same functionality but with better consistency.

Before, the setter raised unconditionally, I think what it meant is to enforce a `XLACheckpointIO`

File: pytorch_lightning/plugins/training_type/deepspeed.py
Comment (carmocca) [reply pytorch_lightning/plugins/training_type/deepspeed.py:880]: Why? This class does not have any custom checkpointing logic

Comment (carmocca) [reply pytorch_lightning/plugins/training_type/deepspeed.py:880]: Mmmm I'm conflicted here.

This plugin doesn't use a checkpoint_io at all. Which might suggest it's good to raise

But the base implementation will set (through the setter) a default `checkpoint_io` which means the exception would always be raised.

I think it's more natural not to override the setter and leave it be. The problem is that this TTP implementation violates the base interface as the base requires the `checkpoint_io` but this subclass does not.

File: pytorch_lightning/plugins/training_type/training_type_plugin.py
Comment (four4fish) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:50]: We need this to be self._precision_plugin instead of direct set to self.precision_plugin
[In 3/n move accelerator into strategy](https://github.com/PyTorchLightning/pytorch-lightning/pull/11022) The accelerator_connector depends on distinguishing is user set up a non_None precision plugin or we set default to decide whether call _select_precision_plugin(), so self._precision_plugin is what user passed in, and self.precision_plugin is default value

Comment (four4fish) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:50]:
```suggestion
        self._checkpoint_io = checkpoint_io
        self._precision_plugin = precision_plugin
```
```suggestion
        self.checkpoint_io = checkpoint_io
        self.precision_plugin = precision_plugin
```

File: pytorch_lightning/plugins/training_type/training_type_plugin.py
Comment (four4fish) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:62]:
```suggestion
    def precision_plugin(self) -> PrecisionPlugin:
        return self._precision_plugin if self._precision_plugin is not None else PrecisionPlugin()
```

File: pytorch_lightning/plugins/training_type/training_type_plugin.py
Comment (four4fish) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:50]: Add a issue to discuss future 
https://github.com/PyTorchLightning/pytorch-lightning/issues/11097

Comment (carmocca) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py]: Do the current changes fix #11097?

File: pytorch_lightning/plugins/training_type/training_type_plugin.py
Comment (four4fish) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:53]:
```suggestion
        self._checkpoint_io = checkpoint_io
        self._precision_plugin = precision_plugin
```

Comment (carmocca) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:53]: Can you explain why? Don't we want the subclasses that override the `setter` with validation (e.g. tpu) to run the validation on init?

See the test I just added in my last commit. It would fail with your proposed change

Comment (four4fish) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:53]: Yeah sure. 
To distinguish user passed Trainer(strategy = CustomStrategy()) Versus. Trainer(stategy = CustomStrategy(precision_plugin = PrecisionPlugin())) in accelerator_connector
If we call strategy.precision_plugin, it will execute the @property which will assign a default value. Then we don't know did user pass in None or a precision_plugin. The current accelerator logic is depends on strategy._precision_plugin to decide whether user passed None and we should choose precision plugin for them. 
This is what I'm trying to say in #11097, we want to distinguish what user passed in vs default value

Could you run tests/accelerator/test_accelerator_test.py? I think with this change, the custom plugin test will fail

Comment (carmocca) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:53]: CI looks green

File: pytorch_lightning/plugins/training_type/tpu_spawn.py
Comment (rohitgr7) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py]: yeah, I'd suggest by default it should use XLACheckpointIO but should allow users to be able to set a custom one too. can be updated in a separate PR.

File: pytorch_lightning/plugins/training_type/deepspeed.py
Comment (rohitgr7) [reply pytorch_lightning/plugins/training_type/deepspeed.py:882]: if we remove this and someone calls plugin.checkpoint_io, it will return `TorchCheckpointIO` I think.
To avoid the conflict with the base one you can check
```py
def checkpoint_io(self, plugin: CheckpointIO)
    if plugin is not None:
        raise MisconfigurationException(...)
```
since it's optional anyways in the base class.

File: pytorch_lightning/plugins/training_type/training_type_plugin.py
Comment (rohitgr7) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:73]: quick ques: why return `TorchCheckpointIO` here if there is no checkpoint_io?

File: pytorch_lightning/plugins/training_type/training_type_plugin.py
Comment (four4fish) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:53]: I see why it worked. in accelerator_connector.py line 406
```
if (
            isinstance(self.strategy, TrainingTypePlugin) and self.strategy._precision_plugin is None
        ) or not isinstance(self.strategy, TrainingTypePlugin):
      call choose precision_plugin logic
```
Because training_type_plugin.precision_plugin haven't been called yet, the training_type_plugin._precision_plugin does not exist, so it guaranteed to be None, so it worked.

But this is VERY fragile, when the training_type_plugin.precision_plugin get called, both `training_type_plugin._precision_plugin` and `training_type_plugin.precision_plugin` will be the default value instead of what user set. If this happens before final_training_type_plugin(), even user themselves called it before pass into trainer, the precision plugin selection logic won't trigger and we will have silent bugs.  

As the test passes for now, feel free to land. We can discuss the longterm robust solution when we solve #11097

File: pytorch_lightning/plugins/training_type/training_type_plugin.py
Comment (carmocca) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:73]: I'm following the same pattern used below for `precision_plugin`

File: pytorch_lightning/plugins/training_type/deepspeed.py
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/deepspeed.py:877]: Just noticed that DeepSpeed overrides `save_checkpoint` and `load_checkpoint`, but not `remove_checkpoint`. Is that correct? Will the removal work with `TorchCheckpointIO`?

File: pytorch_lightning/plugins/training_type/training_type_plugin.py
Comment (carmocca) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:53]: I fully agree. This is not a long-term solution. The accelerator connector needs a revamp

File: pytorch_lightning/plugins/training_type/deepspeed.py
Comment (carmocca) [reply pytorch_lightning/plugins/training_type/deepspeed.py:882]: Done, although I still think we shouln't have these checks. They limit the flexibiltiy.

File: pytorch_lightning/plugins/training_type/training_type_plugin.py
Comment (four4fish) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:50]: Why accelerator here is different?

Comment (carmocca) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:50]: There wasn't a property for it here when I opened the PR. Let's fix it :)

```suggestion
        self.accelerator = accelerator
```

**Review by four4fish — APPROVED** [review]

File: pytorch_lightning/plugins/training_type/deepspeed.py
Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/deepspeed.py:877]: Yes, I think so because it has `recursive=True`

File: pytorch_lightning/plugins/training_type/training_type_plugin.py
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:50]: why are these renamed? You can't have a property and attribute with the same name. What is the actual fix?

Comment (four4fish) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:50]: I was confused about this PR and thought it changes logic too, but seems not. Only logic change is for checkpoint_io.
Before accelerator and precision_plugin in __init__ set self._accelerator/_precision_plugin directly, with changes in this PR it calling setter functions (`self.accelerator` is actually set `self._accelerator` in the accelerator setter method).

File: pytorch_lightning/plugins/training_type/training_type_plugin.py
Comment (four4fish) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:53]: cc @awaelchli for this discussion, which is related to why ._accelerator become .accelerator

File: pytorch_lightning/plugins/training_type/training_type_plugin.py
Comment (carmocca) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:50]: Exactly, also see the changed test

File: pytorch_lightning/plugins/training_type/deepspeed.py
Comment (rohitgr7) [inline pytorch_lightning/plugins/training_type/deepspeed.py:879]: nit: just for consistency with your other changes.
```suggestion
    def checkpoint_io(self, io: Optional[CheckpointIO]) -> None:
        if io is not None:
            raise MisconfigurationException("DeepSpeed does not support custom `CheckpointIO` plugin.")
```

**Review by rohitgr7 — APPROVED** [review]

File: pytorch_lightning/plugins/training_type/deepspeed.py
Comment (tchaton) [reply pytorch_lightning/plugins/training_type/deepspeed.py:877]: Might be better to delete the directory directly. This might be cleaner in `TorchCheckpointIO`.

```py
        fs = get_filesystem(path)
        if fs.exists(path):
            if fs.isdir(path):
                fs.rmdir(path)
            else:
                fs.rm(path, recursive=True)
            log.debug(f"Removed checkpoint: {path}")
```

**Review by tchaton — APPROVED** [review]
LGTM when all comments are addressed

File: pytorch_lightning/plugins/training_type/deepspeed.py
Comment (carmocca) [reply pytorch_lightning/plugins/training_type/deepspeed.py:877]: Rmdir is only for empty directories.

**Review by awaelchli — APPROVED** [review]

**Pull request merged**
